### PR TITLE
[spark] SQL: supports INSERT OVERWRITE syntax

### PR DESCRIPTION
**Pull request by YannByron**
### Purpose

Supports `INSERT OVERWRITE table (PARTITION (...)) ...` in Spark SQL

Linked issue: close #xxx

### CODE REVIEW

File: paimon-spark/paimon-spark-common/src/main/java/org/apache/paimon/spark/SparkTypeUtils.java [hunk 1]
```diff
@@ -173,8 +173,7 @@ public DataType visit(RowType rowType) {
             List<StructField> fields = new ArrayList<>(rowType.getFieldCount());
             for (DataField field : rowType.getFields()) {
                 StructField structField =
-                        DataTypes.createStructField(
-                                field.name(), field.type().accept(this), field.type().isNullable());
+                        DataTypes.createStructField(field.name(), field.type().accept(this), true);
```
Comment (YannByron) [inline paimon-spark/paimon-spark-common/src/main/java/org/apache/paimon/spark/SparkTypeUtils.java:176]: This change is for https://github.com/apache/incubator-paimon/issues/1048. Should submit another pr for this, just leave it for the pr's tests.

File: paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/PaimonCommand.scala [hunk 2]
```diff
@@ -44,4 +47,50 @@ trait PaimonCommand {
...
+      filter: Filter,
+      partitionColumns: Seq[String],
+      resolver: Resolver): Map[String, String] = {
+    splitConjunctiveFilters(filter).map {
+      case EqualNullSafe(attribute, value) =>
+        if (isNestedFilterInValue(value)) {
+          throw new RuntimeException(
+            s"Not support the complex partition value in EqualNullSafe when run `INSERT OVERWRITE`.")
+        } else {
+          partitionColumns.find(resolver(_, attribute)).map((_, value.toString)).getOrElse {
```
Comment (JingsongLi) [inline paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/PaimonCommand.scala:74]: We should use `PredicateBuilder.convertJavaObject` to convert spark java object to our internal data structure.

File: paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/WriteIntoPaimonTable.scala [hunk 3]
```diff
@@ -113,7 +120,9 @@ case class WriteIntoPaimonTable(table: FileStoreTable, overwrite: Boolean, data:
         .map(deserializeCommitMessage(serializer, _))
 
     try {
-      val tableCommit = writeBuilder.newCommit().asInstanceOf[InnerTableCommit]
+      val tableCommit = writeBuilder
+        .withOverwrite(overwritePartition.asJava)
```
Comment (JingsongLi) [inline paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/WriteIntoPaimonTable.scala:124]: if there is no overwrite, we should not invoke `withOverwrite(emptyMap)` here?
I don't understand why tests passing.

File: paimon-spark/paimon-spark-common/src/test/scala/org/apache/paimon/spark/sql/InsertOverwriteSuite.scala [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+ * limitations under the License.
+ */
+package org.apache.paimon.spark.sql
+
+import org.apache.paimon.WriteMode.{APPEND_ONLY, CHANGE_LOG}
+
+import org.apache.spark.sql.Row
+import org.scalactic.source.Position
+
+class InsertOverwriteSuite extends PaimonSparkTestBase {
```
Comment (JingsongLi) [inline paimon-spark/paimon-spark-common/src/test/scala/org/apache/paimon/spark/sql/InsertOverwriteSuite.scala:25]: in our code style, we use xxTest instead of Suite

File: paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/WriteIntoPaimonTable.scala (hunk 3 above)
Comment (YannByron) [reply paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/WriteIntoPaimonTable.scala:124]: if no overwrite here, `overwritePartition` is actually `null`, that is consistent with the behavior of `insert into`.

File: paimon-spark/paimon-spark-common/src/test/scala/org/apache/paimon/spark/sql/InsertOverwriteSuite.scala (hunk 4 above)
Comment (YannByron) [reply paimon-spark/paimon-spark-common/src/test/scala/org/apache/paimon/spark/sql/InsertOverwriteSuite.scala:25]: OK

File: paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/PaimonCommand.scala (hunk 2 above)
Comment (YannByron) [reply paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/PaimonCommand.scala:74]: OK, and i'll add more UTs which use the different data types as the partition columns.

File: paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/WriteIntoPaimonTable.scala (hunk 3 above)
Comment (JingsongLi) [reply paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/WriteIntoPaimonTable.scala:124]: I get it, you are right.

**Review by JingsongLi — APPROVED** [review]
Thanks @YannByron , +1 for this.

**Comment by YannByron** [comment]
This pr just supports `static insert overwrite`, `dynamic overwrite` is under planning.
